File: analysis/diagnostics.py
# Ethan Doughty
# diagnostics.py

from __future__ import annotations

from runtime.shapes import Shape
from ir.ir import *
# ---------------
# Pretty-printing
# ---------------

def pretty_expr_ir(expr: Expr) -> str:
    if isinstance(expr, Var):
        return expr.name
    if isinstance(expr, Const):
        return str(expr.value)
    if isinstance(expr, StringLit):
        return f"'{expr.value}'"
    if isinstance(expr, Neg):
        return f"(-{pretty_expr_ir(expr.operand)})"
    if isinstance(expr, Transpose):
        return pretty_expr_ir(expr.operand) + "'"
    if isinstance(expr, Lambda):
        params = ", ".join(expr.params)
        return f"@({params}) <body>"
    if isinstance(expr, FuncHandle):
        return f"@{expr.name}"
    if isinstance(expr, Apply):
        base = pretty_expr_ir(expr.base) if isinstance(expr.base, Var) else "<expr>"
        args_s = ", ".join(pretty_index_arg_ir(a) for a in expr.args)
        return f"{base}({args_s})"
    if isinstance(expr, CurlyApply):
        base = pretty_expr_ir(expr.base) if isinstance(expr.base, Var) else "<expr>"
        args_s = ", ".join(pretty_index_arg_ir(a) for a in expr.args)
        return f"{base}{{{args_s}}}"
    if isinstance(expr, MatrixLit):
        return "[matrix]"
    if isinstance(expr, CellLit):
        return "{cell}"
    if isinstance(expr, BinOp):
        if expr.op == ":":
            return f"{pretty_expr_ir(expr.left)}:{pretty_expr_ir(expr.right)}"
        return f"({pretty_expr_ir(expr.left)} {expr.op} {pretty_expr_ir(expr.right)})"
    return "<expr>"

def pretty_index_arg_ir(arg: IndexArg) -> str:
    if isinstance(arg, Colon):
        return ":"
    if isinstance(arg, Range):
        return f"{pretty_expr_ir(arg.start)}:{pretty_expr_ir(arg.end)}"
    if isinstance(arg, IndexExpr):
        return pretty_expr_ir(arg.expr)
    return "<idx>"

# ------------------------
# Warning message builders
# ------------------------

def has_unsupported(diags: list[str]) -> bool:
    """Check if any diagnostic is an unsupported statement warning.

    Args:
        diags: List of diagnostic messages

    Returns:
        True if any diagnostic starts with W_UNSUPPORTED_
    """
    return any(d.startswith("W_UNSUPPORTED_") for d in diags)

def warn_reassign_incompatible(line: int, name: str, new_shape: Shape, old_shape: Shape) -> str:
    return (
        f"Line {line}: Variable '{name}' reassigned with incompatible shape "
        f"{new_shape} (previously {old_shape})"
    )

def warn_suspicious_comparison_matrix_scalar(
    line: int, op: str, left_expr: Expr, right_expr: Expr, left: Shape, right: Shape
) -> str:
    return (
        f"Line {line}: Suspicious comparison between matrix and scalar in "
        f"({pretty_expr_ir(left_expr)} {op} {pretty_expr_ir(right_expr)}) ({left} vs {right}). "
        f"In MATLAB this is elementwise and may produce a logical matrix."
    )

def warn_matrix_to_matrix_comparison(
    line: int, op: str, left_expr: Expr, right_expr: Expr, left: Shape, right: Shape
) -> str:
    return (
        f"Line {line}: Matrix-to-matrix comparison in "
        f"({pretty_expr_ir(left_expr)} {op} {pretty_expr_ir(right_expr)}) ({left} vs {right}). "
        f"In MATLAB this is elementwise and may produce a logical matrix."
    )

def warn_logical_op_non_scalar(
    line: int, op: str, left_expr: Expr, right_expr: Expr, left: Shape, right: Shape
) -> str:
    return (
        f"Line {line}: Logical operator {op} used with non-scalar operand(s) in "
        f"({pretty_expr_ir(left_expr)} {op} {pretty_expr_ir(right_expr)}) ({left} vs {right})."
    )

def warn_indexing_scalar(line: int, expr: Expr) -> str:
    return (
        f"Line {line}: Indexing applied to scalar in {pretty_expr_ir(expr)}. "
        f"Treating result as unknown."
    )

def warn_too_many_indices(line: int, expr: Expr) -> str:
    return (
        f"Line {line}: Too many indices for 2D matrix in {pretty_expr_ir(expr)}. "
        f"Treating result as unknown."
    )

def warn_range_endpoints_must_be_scalar(
    line: int, arg: IndexArg, start_shape: Shape, end_shape: Shape
) -> str:
    return (
        f"Line {line}: Range endpoints in indexing must be scalar; got "
        f"{start_shape} and {end_shape} in {pretty_index_arg_ir(arg)}. Treating result as unknown."
    )

def warn_invalid_range_end_lt_start(line: int, arg: IndexArg) -> str:
    return f"Line {line}: Invalid range in indexing ({pretty_index_arg_ir(arg)}): end < start."

def warn_non_scalar_index_arg(line: int, arg: IndexArg, shape: Shape) -> str:
    return (
        f"Line {line}: Non-scalar index argument {pretty_index_arg_ir(arg)} has shape {shape}. "
        f"Treating indexing result as unknown."
    )

def warn_elementwise_mismatch(line: int, op: str, left_expr: Expr, right_expr: Expr, left: Shape, right: Shape) -> str:
    return (
        f"Line {line}: Elementwise {op} mismatch in "
        f"({pretty_expr_ir(left_expr)} {op} {pretty_expr_ir(right_expr)}): {left} vs {right}"
    )

def warn_matmul_mismatch(
    line: int, left_expr: Expr, right_expr: Expr, left: Shape, right: Shape, suggest_elementwise: bool
) -> str:
    msg = (
        f"Line {line}: Dimension mismatch in expression "
        f"({pretty_expr_ir(left_expr)} * {pretty_expr_ir(right_expr)}): "
        f"inner dims {left.cols} vs {right.rows} (shapes {left} and {right})"
    )
    if suggest_elementwise:
        msg += ". Did you mean elementwise multiplication (.*)?"
    return msg


def warn_unsupported_stmt(line: int, raw: str, targets: list[str]) -> str:
    """Warning for unsupported/opaque statement.

    Args:
        line: Source line number
        raw: Original statement text (if available)
        targets: Variables that will be set to unknown

    Returns:
        Warning message string with stable code token W_UNSUPPORTED_STMT
    """
    target_str = ", ".join(targets) if targets else "(none)"
    raw_str = f" '{raw}'" if raw else ""
    return f"W_UNSUPPORTED_STMT line={line} targets={target_str}{raw_str}"

def warn_unknown_function(line: int, name: str) -> str:
    """Warning for unrecognized function call.

    Args:
        line: Source line number
        name: Name of the unrecognized function

    Returns:
        Warning message string with code W_UNKNOWN_FUNCTION
    """
    return f"Line {line}: Function '{name}' is not recognized; treating result as unknown [W_UNKNOWN_FUNCTION]"

def warn_unsupported_multi_assign(line: int) -> str:
    """Warning for unsupported destructuring assignment (Phase A stub).

    Args:
        line: Source line number

    Returns:
        Warning message string with code W_UNSUPPORTED_MULTI_ASSIGN
    """
    return f"W_UNSUPPORTED_MULTI_ASSIGN line {line}: Destructuring assignment not yet supported (Phase C)"

def warn_function_arg_count_mismatch(line: int, func_name: str, expected: int, got: int) -> str:
    """Warning for function called with wrong number of arguments."""
    return f"W_FUNCTION_ARG_COUNT_MISMATCH line {line}: function {func_name} expects {expected} arguments, got {got}"

def warn_recursive_function(line: int, func_name: str) -> str:
    """Warning for recursive function call (not supported, returns unknown)."""
    return f"W_RECURSIVE_FUNCTION line {line}: recursive call to {func_name} not supported (returns unknown)"

def warn_procedure_in_expr(line: int, func_name: str) -> str:
    """Warning for procedure (no return value) used in expression context."""
    return f"W_PROCEDURE_IN_EXPR line {line}: procedure {func_name} has no return value, cannot be used in expression"

def warn_multi_assign_non_call(line: int) -> str:
    """Warning for destructuring assignment with non-function-call RHS."""
    return f"W_MULTI_ASSIGN_NON_CALL line {line}: destructuring assignment requires function call on RHS"

def warn_multi_assign_builtin(line: int, func_name: str) -> str:
    """Warning for destructuring assignment with builtin function (no multi-return)."""
    return f"W_MULTI_ASSIGN_BUILTIN line {line}: builtin {func_name} does not support multiple returns"

def warn_multi_assign_count_mismatch(line: int, func_name: str, expected: int, got: int) -> str:
    """Warning for destructuring assignment target count mismatch."""
    return f"W_MULTI_ASSIGN_COUNT_MISMATCH line {line}: function {func_name} returns {expected} values, got {got} targets"

def warn_string_arithmetic(line: int, op: str, left_shape: Shape, right_shape: Shape) -> str:
    """Warning for invalid string arithmetic (string + matrix/scalar)."""
    return f"W_STRING_ARITHMETIC line {line}: Invalid string arithmetic ({left_shape} {op} {right_shape})"

def warn_struct_field_not_found(line: int, field: str, struct_shape: Shape) -> str:
    """Warning for accessing non-existent struct field."""
    return f"W_STRUCT_FIELD_NOT_FOUND line {line}: Field '{field}' not found in {struct_shape}"

def warn_field_access_non_struct(line: int, base_shape: Shape) -> str:
    """Warning for field access on non-struct value."""
    return f"W_FIELD_ACCESS_NON_STRUCT line {line}: Field access on non-struct value ({base_shape})"

def warn_curly_indexing_non_cell(line: int, base_shape: Shape) -> str:
    """Warning for curly indexing on non-cell value."""
    return f"W_CURLY_INDEXING_NON_CELL line {line}: Curly indexing on non-cell value ({base_shape})"

def warn_cell_assign_non_cell(line: int, base_name: str, base_shape: Shape) -> str:
    """Warning for cell assignment on non-cell variable."""
    return f"W_CELL_ASSIGN_NON_CELL line {line}: Cell assignment to non-cell variable '{base_name}' ({base_shape})"

def warn_return_outside_function(line: int) -> str:
    """Warning for return statement outside function body."""
    return f"W_RETURN_OUTSIDE_FUNCTION line {line}: return statement outside function body"

def warn_break_outside_loop(line: int) -> str:
    """Warning for break statement outside loop (v0.11.1 will validate)."""
    return f"W_BREAK_OUTSIDE_LOOP line {line}: break statement outside loop (treated as no-op)"

def warn_continue_outside_loop(line: int) -> str:
    """Warning for continue statement outside loop (v0.11.1 will validate)."""
    return f"W_CONTINUE_OUTSIDE_LOOP line {line}: continue statement outside loop (treated as no-op)"

def warn_lambda_call_approximate(line: int, var_name: str) -> str:
    """Warning for calling function handle variable (v0.12.0 limitation)."""
    return f"W_LAMBDA_CALL_APPROXIMATE line {line}: Calling function handle '{var_name}' returns unknown (body analysis deferred to v0.12.1)"

def warn_lambda_arg_count_mismatch(line: int, expected: int, got: int) -> str:
    """Warning for lambda called with wrong number of arguments."""
    return f"W_LAMBDA_ARG_COUNT_MISMATCH line {line}: lambda expects {expected} arguments, got {got}"

def warn_recursive_lambda(line: int) -> str:
    """Warning for recursive lambda call (not supported, returns unknown)."""
    return f"W_RECURSIVE_LAMBDA line {line}: recursive lambda call not supported (returns unknown)"

def warn_end_outside_indexing(line: int) -> str:
    """Warning for 'end' keyword used outside indexing context."""
    return f"W_END_OUTSIDE_INDEXING line {line}: 'end' keyword only valid inside indexing expressions"

File: src/Diagnostics.fs
module Diagnostics

open Ir
open Shapes

// ---------------------------------------------------------------------------
// Warning tier classification
// Codes suppressed in default mode; shown only with --strict.
// ---------------------------------------------------------------------------

let STRICT_ONLY_CODES : Set<string> =
    Set.ofList [
        // Parser / recognition limits
        "W_UNSUPPORTED_STMT"; "W_UNSUPPORTED_MULTI_ASSIGN"
        "W_UNKNOWN_FUNCTION"; "W_EXTERNAL_PARSE_ERROR"
        // Cascade-prone field / cell warnings
        "W_STRUCT_FIELD_NOT_FOUND"; "W_FIELD_ACCESS_NON_STRUCT"
        "W_CURLY_INDEXING_NON_CELL"; "W_CELL_ASSIGN_NON_CELL"
        // Informational (not bugs)
        "W_REASSIGN_INCOMPATIBLE"; "W_RECURSIVE_FUNCTION"
        "W_RECURSIVE_LAMBDA"; "W_LAMBDA_CALL_APPROXIMATE"
        "W_MULTI_ASSIGN_NON_CALL"; "W_MULTI_ASSIGN_BUILTIN"
        // Stylistic / suspicious
        "W_SUSPICIOUS_COMPARISON"; "W_MATRIX_COMPARISON"
        "W_LOGICAL_OP_NON_SCALAR"; "W_STRING_ARITHMETIC"
        // 2D shape system limitation
        "W_TOO_MANY_INDICES"
        // cellfun/arrayfun output uniformity
        "W_CELLFUN_NON_UNIFORM"
    ]

/// Coder-mode-only warning codes (emitted only when --coder is active).
let CODER_ONLY_CODES : Set<string> =
    Set.ofList [
        "W_CODER_VARIABLE_SIZE"
        "W_CODER_CELL_ARRAY"
        "W_CODER_DYNAMIC_FIELD"
        "W_CODER_TRY_CATCH"
        "W_CODER_UNSUPPORTED_BUILTIN"
        "W_CODER_RECURSION"
    ]

// ---------------------------------------------------------------------------
// Diagnostic record
// ---------------------------------------------------------------------------

type Diagnostic = {
    line:       int
    code:       string   // W_* prefix
    message:    string
    relatedLine: int option
    col:        int
    relatedCol: int option
}

let diagnosticToString (d: Diagnostic) : string =
    // Special case: W_UNSUPPORTED_STMT uses line= format
    if d.code = "W_UNSUPPORTED_STMT" then
        d.code + " line=" + string d.line + " " + d.message
    elif d.code <> "" then
        d.code + " line " + string d.line + ": " + d.message
    else
        "Line " + string d.line + ": " + d.message

// ---------------------------------------------------------------------------
// Pretty-printing helpers for IR expressions
// ---------------------------------------------------------------------------

let rec prettyExprIr (expr: Expr) : string =
    match expr with
    | Var(_, name)        -> name
    | Const(_, v)         ->
        if v = System.Math.Floor(v : float) && not (System.Double.IsInfinity v) then
            string (int64 v)
        else
            string v
    | StringLit(_, s)     -> "'" + s + "'"
    | Neg(_, operand)     -> "(-" + prettyExprIr operand + ")"
    | Not(_, operand)     -> "(~" + prettyExprIr operand + ")"
    | Transpose(_, operand) -> prettyExprIr operand + "'"
    | Lambda(_, parms, _) ->
        let ps = parms |> String.concat ", "
        "@(" + ps + ") <body>"
    | FuncHandle(_, name) -> "@" + name
    | Apply(_, base_, args) ->
        let baseName =
            match base_ with
            | Var(_, n) -> n
            | _         -> "<expr>"
        let argsStr = args |> List.map prettyIndexArgIr |> String.concat ", "
        baseName + "(" + argsStr + ")"
    | CurlyApply(_, base_, args) ->
        let baseName =
            match base_ with
            | Var(_, n) -> n
            | _         -> "<expr>"
        let argsStr = args |> List.map prettyIndexArgIr |> String.concat ", "
        baseName + "{" + argsStr + "}"
    | MatrixLit _            -> "[matrix]"
    | CellLit _              -> "{cell}"
    | BinOp(_, op, left, right) ->
        if op = ":" then
            prettyExprIr left + ":" + prettyExprIr right
        else
            "(" + prettyExprIr left + " " + op + " " + prettyExprIr right + ")"
    | FieldAccess(_, base_, field) ->
        prettyExprIr base_ + "." + field
    | End _ -> "end"

and prettyIndexArgIr (arg: IndexArg) : string =
    match arg with
    | Colon _              -> ":"
    | Ir.Range(_, s, e)    -> prettyExprIr s + ":" + prettyExprIr e
    | IndexExpr(_, e)      -> prettyExprIr e

// ---------------------------------------------------------------------------
// Warning message builders
// ---------------------------------------------------------------------------

let makeDiag line code message =
    { line = line; code = code; message = message; relatedLine = None; col = 0; relatedCol = None }

let makeDiagRel line code message relLine =
    { line = line; code = code; message = message; relatedLine = Some relLine; col = 0; relatedCol = None }

let hasUnsupported (diags: Diagnostic list) : bool =
    diags |> List.exists (fun d -> d.code.StartsWith("W_UNSUPPORTED_"))

let warnReassignIncompatible (line: int) (name: string) (newShape: Shape) (oldShape: Shape) : Diagnostic =
    makeDiag line "W_REASSIGN_INCOMPATIBLE"
        ("Variable '" + name + "' reassigned with incompatible shape " + shapeToString newShape +
         " (previously " + shapeToString oldShape + ")")

let warnSuspiciousComparisonMatrixScalar
    (line: int) (op: string) (leftExpr: Expr) (rightExpr: Expr) (left: Shape) (right: Shape) : Diagnostic =
    makeDiag line "W_SUSPICIOUS_COMPARISON"
        ("Suspicious comparison between matrix and scalar in (" +
         prettyExprIr leftExpr + " " + op + " " + prettyExprIr rightExpr +
         ") (" + shapeToString left + " vs " + shapeToString right +
         "). In MATLAB this is elementwise and may produce a logical matrix.")

let warnMatrixToMatrixComparison
    (line: int) (op: string) (leftExpr: Expr) (rightExpr: Expr) (left: Shape) (right: Shape) : Diagnostic =
    makeDiag line "W_MATRIX_COMPARISON"
        ("Matrix-to-matrix comparison in (" +
         prettyExprIr leftExpr + " " + op + " " + prettyExprIr rightExpr +
         ") (" + shapeToString left + " vs " + shapeToString right +
         "). In MATLAB this is elementwise and may produce a logical matrix.")

let warnLogicalOpNonScalar
    (line: int) (op: string) (leftExpr: Expr) (rightExpr: Expr) (left: Shape) (right: Shape) : Diagnostic =
    makeDiag line "W_LOGICAL_OP_NON_SCALAR"
        ("Logical operator " + op + " used with non-scalar operand(s) in (" +
         prettyExprIr leftExpr + " " + op + " " + prettyExprIr rightExpr +
         ") (" + shapeToString left + " vs " + shapeToString right + ").")

let warnIndexingScalar (line: int) (expr: Expr) : Diagnostic =
    makeDiag line "W_INDEXING_SCALAR"
        ("Indexing applied to scalar in " + prettyExprIr expr + ". Treating result as unknown.")

let warnTooManyIndices (line: int) (expr: Expr) : Diagnostic =
    makeDiag line "W_TOO_MANY_INDICES"
        ("Too many indices for 2D matrix in " + prettyExprIr expr + ". Treating result as unknown.")

let warnRangeEndpointsMustBeScalar (line: int) (arg: IndexArg) (startShape: Shape) (endShape: Shape) : Diagnostic =
    makeDiag line "W_RANGE_NON_SCALAR"
        ("Range endpoints in indexing must be scalar; got " +
         shapeToString startShape + " and " + shapeToString endShape +
         " in " + prettyIndexArgIr arg + ". Treating result as unknown.")

let warnInvalidRangeEndLtStart (line: int) (arg: IndexArg) : Diagnostic =
    makeDiag line "W_INVALID_RANGE"
        ("Invalid range in indexing (" + prettyIndexArgIr arg + "): end < start.")

let warnNonScalarIndexArg (line: int) (arg: IndexArg) (shape: Shape) : Diagnostic =
    makeDiag line "W_NON_SCALAR_INDEX"
        ("Non-scalar index argument " + prettyIndexArgIr arg + " has shape " + shapeToString shape +
         ". Treating indexing result as unknown.")

let warnElementwiseMismatch
    (line: int) (op: string) (leftExpr: Expr) (rightExpr: Expr) (left: Shape) (right: Shape) : Diagnostic =
    makeDiag line "W_ELEMENTWISE_MISMATCH"
        ("Elementwise " + op + " mismatch in (" +
         prettyExprIr leftExpr + " " + op + " " + prettyExprIr rightExpr +
         "): " + shapeToString left + " vs " + shapeToString right)

let warnMatmulMismatch
    (line: int) (leftExpr: Expr) (rightExpr: Expr) (left: Shape) (right: Shape) (suggestElementwise: bool) : Diagnostic =
    let leftCols =
        match left with Matrix(_, c) -> dimStr c | _ -> "?"
    let rightRows =
        match right with Matrix(r, _) -> dimStr r | _ -> "?"
    let mutable msg =
        "Dimension mismatch in expression (" +
        prettyExprIr leftExpr + " * " + prettyExprIr rightExpr + "): inner dims " +
        leftCols + " vs " + rightRows +
        " (shapes " + shapeToString left + " and " + shapeToString right + ")"
    if suggestElementwise then
        msg <- msg + ". Did you mean elementwise multiplication (.*)?"
    makeDiag line "W_INNER_DIM_MISMATCH" msg

let warnUnsupportedStmt (line: int) (raw: string) (targets: string list) : Diagnostic =
    let targetStr = if targets.IsEmpty then "(none)" else targets |> String.concat ", "
    let rawStr    = if raw = "" then "" else " '" + raw + "'"
    makeDiag line "W_UNSUPPORTED_STMT" ("targets=" + targetStr + rawStr)

let warnUnknownFunction (line: int) (name: string) : Diagnostic =
    makeDiag line "W_UNKNOWN_FUNCTION"
        ("Function '" + name + "' is not recognized; treating result as unknown")

let warnUnsupportedMultiAssign (line: int) : Diagnostic =
    makeDiag line "W_UNSUPPORTED_MULTI_ASSIGN"
        "Destructuring assignment not yet supported (Phase C)"

let warnFunctionArgCountMismatch (line: int) (funcName: string) (expected: int) (got: int) : Diagnostic =
    makeDiag line "W_FUNCTION_ARG_COUNT_MISMATCH"
        ("function " + funcName + " expects " + string expected + " arguments, got " + string got)

let warnRecursiveFunction (line: int) (funcName: string) : Diagnostic =
    makeDiag line "W_RECURSIVE_FUNCTION"
        ("recursive call to " + funcName + " not supported (returns unknown)")

let warnProcedureInExpr (line: int) (funcName: string) : Diagnostic =
    makeDiag line "W_PROCEDURE_IN_EXPR"
        ("procedure " + funcName + " has no return value, cannot be used in expression")

let warnMultiAssignNonCall (line: int) : Diagnostic =
    makeDiag line "W_MULTI_ASSIGN_NON_CALL"
        "destructuring assignment requires function call on RHS"

let warnMultiAssignBuiltin (line: int) (funcName: string) : Diagnostic =
    makeDiag line "W_MULTI_ASSIGN_BUILTIN"
        ("builtin " + funcName + " does not support multiple returns")

let warnMultiAssignCountMismatch (line: int) (funcName: string) (expected: int) (got: int) : Diagnostic =
    makeDiag line "W_MULTI_ASSIGN_COUNT_MISMATCH"
        ("function " + funcName + " returns " + string expected + " values, got " + string got + " targets")

let warnMultiReturnCount (line: int) (fname: string) (supported: string) (got: int) : Diagnostic =
    makeDiag line "W_MULTI_ASSIGN_COUNT_MISMATCH"
        ("builtin " + fname + " supports " + supported + " return values, got " + string got)

let warnStringArithmetic (line: int) (op: string) (leftShape: Shape) (rightShape: Shape) : Diagnostic =
    makeDiag line "W_STRING_ARITHMETIC"
        ("Invalid string arithmetic (" + shapeToString leftShape + " " + op + " " + shapeToString rightShape + ")")

let warnStructFieldNotFound (line: int) (field: string) (structShape: Shape) : Diagnostic =
    makeDiag line "W_STRUCT_FIELD_NOT_FOUND"
        ("Field '" + field + "' not found in " + shapeToString structShape)

let warnFieldAccessNonStruct (line: int) (baseShape: Shape) : Diagnostic =
    makeDiag line "W_FIELD_ACCESS_NON_STRUCT"
        ("Field access on non-struct value (" + shapeToString baseShape + ")")

let warnCurlyIndexingNonCell (line: int) (baseShape: Shape) : Diagnostic =
    makeDiag line "W_CURLY_INDEXING_NON_CELL"
        ("Curly indexing on non-cell value (" + shapeToString baseShape + ")")

let warnCellAssignNonCell (line: int) (baseName: string) (baseShape: Shape) : Diagnostic =
    makeDiag line "W_CELL_ASSIGN_NON_CELL"
        ("Cell assignment to non-cell variable '" + baseName + "' (" + shapeToString baseShape + ")")

let warnIndexAssignTypeMismatch (line: int) (baseName: string) (baseShape: Shape) : Diagnostic =
    makeDiag line "W_INDEX_ASSIGN_TYPE_MISMATCH"
        ("Indexed assignment to non-indexable variable '" + baseName + "' (" + shapeToString baseShape + ")")

let warnReturnOutsideFunction (line: int) : Diagnostic =
    makeDiag line "W_RETURN_OUTSIDE_FUNCTION"
        "return statement outside function body"

let warnBreakOutsideLoop (line: int) : Diagnostic =
    makeDiag line "W_BREAK_OUTSIDE_LOOP"
        "break statement outside loop (treated as no-op)"

let warnContinueOutsideLoop (line: int) : Diagnostic =
    makeDiag line "W_CONTINUE_OUTSIDE_LOOP"
        "continue statement outside loop (treated as no-op)"

let warnLambdaCallApproximate (line: int) (varName: string) : Diagnostic =
    makeDiag line "W_LAMBDA_CALL_APPROXIMATE"
        ("Calling function handle '" + varName + "' returns unknown (body analysis deferred to v0.12.1)")

let warnLambdaArgCountMismatch (line: int) (expected: int) (got: int) : Diagnostic =
    makeDiag line "W_LAMBDA_ARG_COUNT_MISMATCH"
        ("lambda expects " + string expected + " arguments, got " + string got)

let warnRecursiveLambda (line: int) : Diagnostic =
    makeDiag line "W_RECURSIVE_LAMBDA"
        "recursive lambda call not supported (returns unknown)"

let warnEndOutsideIndexing (line: int) : Diagnostic =
    makeDiag line "W_END_OUTSIDE_INDEXING"
        "'end' keyword only valid inside indexing expressions"

let warnExternalParseError (line: int) (fname: string) (sourcePath: string) : Diagnostic =
    makeDiag line "W_EXTERNAL_PARSE_ERROR"
        ("Cannot analyze " + fname + " (parse error in " + sourcePath + "); treating result as unknown")

let warnConstraintConflict (line: int) (varName: string) (value: int) (otherDim: string) (sourceLine: int) : Diagnostic =
    makeDiag line "W_CONSTRAINT_CONFLICT"
        (varName + "=" + string value + " conflicts with " + varName + "==" + otherDim + " (from line " + string sourceLine + ")")

let warnReshapeMismatch (line: int) (inputShape: Shape) (m: string) (n: string) : Diagnostic =
    makeDiag line "W_RESHAPE_MISMATCH"
        ("reshape changes element count: " + shapeToString inputShape +
         " has different element count than " + m + "x" + n)

let warnDivisionByZero (line: int) (leftExpr: Expr) (rightExpr: Expr) : Diagnostic =
    // leftExpr and rightExpr kept for future use / symmetry with Python
    ignore leftExpr
    ignore rightExpr
    makeDiag line "W_DIVISION_BY_ZERO"
        "division by zero: divisor is definitely zero"

let warnIndexOutOfBounds (line: int) (indexVal: string) (dimSize: string) (definite: bool) : Diagnostic =
    let verb = if definite then "exceeds" else "may exceed"
    makeDiag line "W_INDEX_OUT_OF_BOUNDS"
        ("index out of bounds: index range " + indexVal + " " + verb + " dimension " + dimSize)

let warnPossiblyNegativeDim (line: int) (dimVal: string) : Diagnostic =
    makeDiag line "W_POSSIBLY_NEGATIVE_DIM"
        ("non-positive dimension: " + dimVal)

let warnArithmeticTypeMismatch
    (line: int) (op: string) (leftExpr: Expr) (rightExpr: Expr) (left: Shape) (right: Shape) : Diagnostic =
    ignore leftExpr
    ignore rightExpr
    makeDiag line "W_ARITHMETIC_TYPE_MISMATCH"
        ("Arithmetic operator " + op + " requires numeric operands, got " +
         shapeToString left + " and " + shapeToString right)

let warnTransposeTypeMismatch (line: int) (shape: Shape) : Diagnostic =
    makeDiag line "W_TRANSPOSE_TYPE_MISMATCH"
        ("Transpose requires numeric operand, got " + shapeToString shape)

let warnNegateTypeMismatch (line: int) (shape: Shape) : Diagnostic =
    makeDiag line "W_NEGATE_TYPE_MISMATCH"
        ("Negation requires numeric operand, got " + shapeToString shape)

let warnNotTypeMismatch (line: int) (shape: Shape) : Diagnostic =
    makeDiag line "W_NOT_TYPE_MISMATCH"
        ("Logical NOT requires numeric operand, got " + shapeToString shape)

let warnConcatTypeMismatch (line: int) (shape: Shape) : Diagnostic =
    makeDiag line "W_CONCAT_TYPE_MISMATCH"
        ("Concatenation requires numeric elements, got " + shapeToString shape)

let warnMldivideDimMismatch
    (line: int) (leftExpr: Expr) (rightExpr: Expr) (left: Shape) (right: Shape) : Diagnostic =
    let leftRows  = match left  with Matrix(r, _) -> dimStr r | _ -> "?"
    let rightRows = match right with Matrix(r, _) -> dimStr r | _ -> "?"
    makeDiag line "W_MLDIVIDE_DIM_MISMATCH"
        ("Dimension mismatch in mldivide (" +
         prettyExprIr leftExpr + " \\ " + prettyExprIr rightExpr +
         "): A has " + leftRows + " rows but b has " + rightRows +
         " rows (shapes " + shapeToString left + " and " + shapeToString right + ")")

let warnMatrixPowerNonSquare (line: int) (expr: Expr) (shape: Shape) : Diagnostic =
    let rows = match shape with Matrix(r, _) -> dimStr r | _ -> "?"
    let cols = match shape with Matrix(_, c) -> dimStr c | _ -> "?"
    makeDiag line "W_MATRIX_POWER_NON_SQUARE"
        ("Matrix power (^) requires square matrix; " +
         prettyExprIr expr + " has shape " + shapeToString shape +
         " (" + rows + " rows, " + cols + " cols)")

let warnHorzcatRowMismatch (line: int) (rowA: Dim) (rowB: Dim) : Diagnostic =
    makeDiag line "W_HORZCAT_ROW_MISMATCH"
        ("Horizontal concatenation row count mismatch: " + dimStr rowA + " vs " + dimStr rowB)

let warnVertcatColMismatch (line: int) (colA: Dim) (colB: Dim) : Diagnostic =
    makeDiag line "W_VERTCAT_COL_MISMATCH"
        ("Vertical concatenation column count mismatch: " + dimStr colA + " vs " + dimStr colB)

let warnCellfunNonUniform (line: int) : Diagnostic =
    makeDiag line "W_CELLFUN_NON_UNIFORM"
        "cellfun returns non-scalar elements; set 'UniformOutput',false to collect into a cell array"

// ---------------------------------------------------------------------------
// Coder-mode warning builders (W_CODER_* family)
// ---------------------------------------------------------------------------

let warnCoderVariableSize (line: int) (varName: string) (shape: Shape) : Diagnostic =
    makeDiag line "W_CODER_VARIABLE_SIZE"
        ("Variable '" + varName + "' has unbounded dimension " + shapeToString shape +
         " (MATLAB Coder requires fixed-size or coder.varsize declaration)")

let warnCoderCellArray (line: int) (varName: string) : Diagnostic =
    makeDiag line "W_CODER_CELL_ARRAY"
        ("Variable '" + varName + "' is a cell array (limited support in MATLAB Coder)")

let warnCoderDynamicField (line: int) : Diagnostic =
    makeDiag line "W_CODER_DYNAMIC_FIELD"
        "Dynamic struct field access s.(expr) is not supported by MATLAB Coder"

let warnCoderTryCatch (line: int) : Diagnostic =
    makeDiag line "W_CODER_TRY_CATCH"
        "try/catch is not supported by MATLAB Coder"

let warnCoderUnsupportedBuiltin (line: int) (fname: string) : Diagnostic =
    makeDiag line "W_CODER_UNSUPPORTED_BUILTIN"
        ("Builtin '" + fname + "' is not supported by MATLAB Coder")

let warnCoderRecursion (line: int) (fname: string) : Diagnostic =
    makeDiag line "W_CODER_RECURSION"
        ("Recursive call to '" + fname + "' (MATLAB Coder supports limited recursion)")
